### ISSUE TITLE
S4049 should not raise on return type IEnumerable<T>

### ISSUE THREAD

**Issue by Corniel** [SonarSource/sonar-dotnet#3143]
### Description
The documentation on S4049 explains that `T[]` is excluded. I would argue that `IEnumerable<T>` (and implementations of `IEnumerable<T>` should also be return types that should not raise S4049.

Specially LINQ queries can be _expensive_, and it is definitely more that just returning a value type or a pointer to an object (which is the reasoning behind S4049).

### Repro steps
``` C#
public class S4049_IEnumerable
{
    public IEnumerable<char> GetChars() => "some characters"; // not compliant, defacto a string.

    public IEnumerable<char> GetCharsAsEnumerable() => "some characters".Where(ch => char.IsLetter(ch)); // compliant?!

    public IEnumerable<int> GetNumbers() => new[] { 1, 2 }.Where(n => (n % 2) == 1); // compliant.

    public List<string> GetStrings() => new List<string> { "hello", "world" }; // compliant.

    public List<int> GetNumberList() => GetNumbers().ToList(); // compliant.
}
```
### Related information
* SonarC# Version 8.3.0.14607

**Comment by andrei-epure-sonarsource** [comment]
hey @Corniel 

>  The documentation on S4049 explains that T[] is excluded. 

Indeed, `and returns a value that is not an array.` (https://jira.sonarsource.com/browse/RSPEC-4049)

I believe too that methods should be used for expensive calls instead of properties. However, how do you see the cases of 

```
public class
{
  private List<string> customers;
  public Dictionary<string, string> customerDictionary;

  public IEnumerable<string> GetCustomers => customers;

  public IEnumerable<string> GetCustomers2 => customerDictionary.Keys;

}
```

**Comment by Corniel** [comment]
``` C#
public class
{
  private List<string> customers;
  public Dictionary<string, string> customerDictionary;

  public IEnumerable<string> GetCustomers() => customers;

  public IEnumerable<string> GetCustomers2() => customerDictionary.Keys;

  public List<string> CustomersAsList => customers;

  public IReadOnlyCollection<string> CustomersAsList => customerDictionary.Keys;
}
```

@andrei-epure-sonarsource I think that both should be allowed to be methods. And that is because of the IEnumerable<T>. The other two added examples explicitly share the **whole** collections, via already instantiated properties, but with IEnumerable you say (via you contract): there are some items, you might want to pick (some of them), and we might need to do things first before we can give them to you.

**Comment by andrei-epure-sonarsource** [comment]
Well, all properties allow to "do things first before we can give them to you". And this is a debatable convention rule - not everyone would consider it useful.

And one can have an IEnumerable property just because the respective property needs to be enumerated upon.

What I understand is that you say "given that LINQ queries are expensive, and IEnumerable comes a lot with LINQ queries, this rule should tolerate IEnumerable because it's better to have LINQ queries in methods".

**Comment by andrei-epure-sonarsource** [comment]
we talked a bit internally and we decided that we 'Won't fix' this one because:
- having `IEnumerable` doesn't mean necessarily you want to process things before returning, it can also mean "I don't want to expose a higher-level abstraction if all you need is to iterate on what I give you"
- this rule is highly debatable and quite noisy, and not in SonarWay
- there is a bigger problem with this rule: the fact that we tell people to move big methods (no matter how big and complex) to properties - so I opened #3200 to track the bigger problem instead

Thanks a lot for this discussion !
